typing/typecore.ml

```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
     -> final_subexpression e
   | _ -> exp
 
+let is_prim ~name funct =
+  match funct.exp_desc with
+  | Texp_ident (_, _, {val_kind=Val_prim{Primitive.prim_name; _}}) ->
+      prim_name = name
+  | _ -> false
+
+(* List labels in a function type, and whether return type is a variable *)
+let rec list_labels_aux env visited ls ty_fun =
+  let ty = expand_head env ty_fun in
+  if TypeSet.mem ty visited then
+    List.rev ls, false
+  else match get_desc ty with
+    | Tarrow (l, _, ty_res, _) ->
+        list_labels_aux env (TypeSet.add ty visited) (l::ls) ty_res
+    | _ ->
+        List.rev ls, is_Tvar ty
+
+let list_labels env ty =
+  let snap = Btype.snapshot () in
+  let result =
+    wrap_trace_gadt_instances env (list_labels_aux env TypeSet.empty []) ty
+  in
+  Btype.backtrack snap;
+  result
+
+
+(* Collecting arguments for function applications *)
+
+type untyped_apply_arg =
+  | Known_arg of
+      {
+        sarg : Parsetree.expression;
+        ty_arg : type_expr;
+        ty_arg0 : type_expr;
+        wrapped_in_some : bool;
+      }
+  | Unknown_arg of
+      {
+        sarg : Parsetree.expression;
+        ty_arg : type_expr;
+      }
+  | Eliminated_optional_arg of
+      {
+        ty_arg : type_expr;
+        level: int;
+      }
+
+type untyped_omitted_param =
+  {
+    ty_arg : type_expr;
+    level: int;
+  }
+
+let remaining_function_type ty_ret rev_args =
```

Comment:
For reviewers: this corresponds to `result_type` at the beginning of `type_application` in the trunk implementation, except that it takes the full (rev)list of `apply_arg`, not just omitted and eliminated arguments.

Reply:
> Would it not be better to reserve Arg for actual arguments, and put erased arguments on the Omitted side.

It is much more important to distinguish omitted parameters -- which cause a closure to be created rather than a function to be called -- from ordinary arguments. Whist `untyped_omitted_param` is local to `type_application` the use of the `arg_or_omitted` is not: the distinction between actual arguments and omitted parameters is preserved all the way through to `translcore`. Getting this distinction clear and correct was a key motivation for this patch (it matters *a lot* for our work on modes like `local`).

Reply:
Indeed, `arg_or_omitted` is kept.
It looks like I got confused by looking simultaneously at the original and new code.

Reply:
Sorry, actually this is not what I wanted to say.
Indeed, `arg_or_omitted` is kept. But `untyped_apply_arg` is not.
And I am still under the impression that, until they get typed, the code would be simpler with eliminated optional arguments on the `Omitted` side. In particular in this function, the point being that there are two kinds of omitted arguments, those that get eliminated and those that are kept, but it only matters later on. Of course, you need to move them to the `Arg` side when you create the concrete argument. Either in `type_apply_arg` or `type_omitted_parameters`; I would actually prefer the later, and it would even give you a justification for the name.

Reply:
Having the meaning of `Omitted` be inconsistent between different parts of the code just makes it much harder to understand for no obvious benefit. Eliminated optional parameters are fundamentally *not* omitted arguments and it is confusing to label them as such.

Reply:
But they are omitted. Seeing them otherwise makes this function hard to understand.
Anyway, I don't want to make this a requirement (this is why I approved).